homeassistant/components/humidifier/__init__.py

```diff
@@ -0,0 +1,288 @@
+    SERVICE_SET_HUMIDIFIER_MODE,
+    SERVICE_SET_PRESET_MODE,
+    SUPPORT_FAN_MODE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_HUMIDITY,
+)
+
+DEFAULT_MIN_HUMIDITY = 30
+DEFAULT_MAX_HUMIDITY = 99
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+SCAN_INTERVAL = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+SET_FAN_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_FAN_MODE): cv.string}
+)
+SET_PRESET_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_PRESET_MODE): cv.string}
+)
+SET_HUMIDIFIER_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_HUMIDIFIER_MODE): vol.In(HUMIDIFIER_MODES)}
+)
+SET_HUMIDITY_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_HUMIDITY): vol.Coerce(float)}
+)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up humidifier devices."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_TURN_ON, ENTITY_SERVICE_SCHEMA, "async_turn_on"
+    )
+    component.async_register_entity_service(
+        SERVICE_TURN_OFF, ENTITY_SERVICE_SCHEMA, "async_turn_off"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_FAN_MODE, SET_FAN_MODE_SCHEMA, "async_set_fan_mode"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_HUMIDIFIER_MODE,
+        SET_HUMIDIFIER_MODE_SCHEMA,
+        "async_set_humidifier_mode",
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_PRESET_MODE, SET_PRESET_MODE_SCHEMA, "async_set_preset_mode"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_HUMIDITY, SET_HUMIDITY_SCHEMA, "async_set_humidity"
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry):
+    """Set up a config entry."""
+    return await hass.data[DOMAIN].async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry):
+    """Unload a config entry."""
+    return await hass.data[DOMAIN].async_unload_entry(entry)
+
+
+class HumidifierDevice(Entity):
+    """Representation of a humidifier device."""
+
+    @property
+    def state(self) -> str:
+        """Return the current state."""
+        return self.humidifier_mode
+
+    @property
+    def state_attributes(self) -> Dict[str, Any]:
+        """Return the optional state attributes."""
+        supported_features = self.supported_features
+        data = {
+            ATTR_HUMIDIFIER_MODES: self.humidifier_modes,
+            ATTR_CURRENT_HUMIDITY: self.current_humidity,
+            ATTR_MIN_HUMIDITY: self.min_humidity,
+            ATTR_MAX_HUMIDITY: self.max_humidity,
+        }
+
+        if supported_features & SUPPORT_TARGET_HUMIDITY:
+            data[ATTR_HUMIDITY] = self.target_humidity
+            data[ATTR_MIN_HUMIDITY] = self.min_humidity
```

Comment:
As I suggested in the const file, we should change it to `target_humidity_low` and `high`. `min` and `max` should still preserved to present the device's min and max humidity range.

Reply:
That is an error, it should be removed from here, we set ATTR_MIN_HUMIDITY and ATTR_MAX_HUMIDITY above already.